src/Stack/Build/ConstructPlan.hs

```diff
@@ -192,6 +188,16 @@ constructPlan mbp0 baseConfigOpts0 locals extraToBuild0 localDumpPkgs loadPackag
     -- or local packages.
     toolMap = getToolMap mbp0
 
+withLoadPackageVersions
+    :: (MonadIO m, HasHttpManager env, MonadReader env m, MonadBaseControl IO m, MonadCatch m, MonadLogger m, HasConfig env)
+    => ((PackageName -> IO (Set Version)) -> m a)
+    -> m a
+withLoadPackageVersions inner =
+    join $ runOnce $ do
```

Comment:
Hmm, I think there's a misunderstanding here about `runOnce`.  `join . runOnce` is essentially `id`.  The action returned by `runOnce` lets us ask for the value.  That action can be run multiple times, but the computation only gets run once (hence the name).